src/prefect/cli/cloud.py

```diff
@@ -35,6 +47,66 @@
 app.add_typer(cloud_app)
 
 
+# Set up a little API server for browser based `prefect cloud login`
+
+login_api = FastAPI()
+
+login_api.add_middleware(
+    CORSMiddleware,
+    allow_origins=["*"],
+    allow_methods=["*"],
+    allow_headers=["*"],
+)
+
+
+class LoginSuccess(BaseModel):
+    api_key: str
+    workspace_id: UUID
+    account_id: UUID
+
+
+class LoginFailed(BaseModel):
+    reason: str
+
+
+class LoginResult(BaseModel):
+    type: Literal["success", "failure"]
+    content: Union[LoginSuccess, LoginFailed]
+
+
+class ServerExit(Exception):
+    pass
+
+
+@login_api.post("/success")
+def receive_login(payload: LoginSuccess):
+    login_api.extra["result"] = LoginResult(type="success", content=payload)
+    login_api.extra["result-event"].set()
+
+
+@login_api.post("/failure")
+def receive_failure(payload: LoginFailed):
+    login_api.extra["result"] = LoginResult(type="failure", content=payload)
+    login_api.extra["result-event"].set()
+
+
+async def serve_login_api(cancel_scope, task_status):
+    task_status.started()
+
+    config = uvicorn.Config(
+        "prefect.cli.login:login_api", port=3001, log_level="critical"
+    )
+    server = uvicorn.Server(config)
+
+    try:
+        await server.serve()
+    except anyio.get_cancelled_exc_class():
+        pass  # Already cancelled, do not cancel again
+    else:
+        # Uvicorn overrides signal handlers so without this Ctrl-C is broken
+        cancel_scope.cancel()
```

Comment:
we tear down the server in each instance, correct?

Reply:
Yeah the server should always tear down. This is really awkward due to some upstream issues.